source/block-reconciler.ts

```diff
@@ -1,28 +1,25 @@
 import { Block } from "./models/block";
 import { BlockHistory } from "./models/block-history";
-import { List as ImmutableList } from "immutable";
 
-type GetBlockByHash<TBlock> = (hash: string) => Promise<TBlock|null>;
+type GetBlockByHash<TBlock> = (hash: string) => Promise<TBlock | null>;
 
 export const reconcileBlockHistory = async <TBlock extends Block>(
 	getBlockByHash: GetBlockByHash<TBlock>,
-	blockHistory: BlockHistory<TBlock>|Promise<BlockHistory<TBlock>>,
+	blockHistory: BlockHistory<TBlock> | Promise<BlockHistory<TBlock>>,
 	newBlock: TBlock,
 	onBlockAdded: (block: TBlock) => Promise<void>,
 	onBlockRemoved: (block: TBlock) => Promise<void>,
 	blockRetention: number = 100,
 ): Promise<BlockHistory<TBlock>> => {
 	blockHistory = await blockHistory;
-	if (isFirstBlock(blockHistory))
-		return await addNewHeadBlock(blockHistory, newBlock, onBlockAdded, blockRetention);
+	if (isFirstBlock(blockHistory)) return await addNewHeadBlock(blockHistory, newBlock, onBlockAdded, blockRetention);
 
 	if (isOlderThanOldestBlock(blockHistory, newBlock)) {
 		blockHistory = await rollback(blockHistory, onBlockRemoved);
 		return await addNewHeadBlock(blockHistory, newBlock, onBlockAdded, blockRetention);
 	}
 
-	if (isAlreadyInHistory(blockHistory, newBlock))
-		return blockHistory;
+	if (isAlreadyInHistory(blockHistory, newBlock)) return blockHistory;
```

Comment:
I prefer to put early returns on their own line.  The reason for this is that it makes it easier when scanning the code to see all of the exit points for the function when they any exit point will be one a line starting with `return`.  This way you can scan the left column (following indents) to quickly locate all function exit points.  This does differ slightly from preconditions, where the `throw` goes on the same line as the `if`.  The reason for this is because the reader should be largely ignoring those lines, they are there mainly to help the typechecker and if you hit them it is (in most cases) a programming error, so someone trying to understand the flow of the code when reading can largely ignore preconditions.  In other projects I have improved on this by creating a `require(condition, message)` function that bundles up the `if (condition) throw new Error(message)` into a single call that makes it clear when reading the start of the line that this is a precondition.  I wouldn't be against applying that throughout this codebase in order to get rid of the `if (condition) throw` and `if(condition) return` inconsistency with when to use newlines.

Reply:
One way to get the `return` on a new-line, is to add the brackets for the conditional.

Reply:
Yeah, I have worked on projects that do that and it does help in satisfying some linters, it also (IMO) results in a bunch of unnecessary noise.  This is similar to semicolons in that there is already enough information to handle proper control flow and all modern editors will do the right thing with regards to formatting so a human doesn't get confused.  Adding `{}` ends up being redundant.